ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp

```diff
@@ -0,0 +1,119 @@
+#include "moreh_clip_grad_norm_step3/device/moreh_clip_grad_norm_step3_device_operation.hpp"
+#include "ttnn/cpp/ttnn/operations/eltwise/binary/binary.hpp"
+#include "ttnn/cpp/ttnn/operations/eltwise/binary/binary_composite.hpp"
+#include "ttnn/operations/creation.hpp"
+#include "ttnn/tensor/shape/shape.hpp"
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::moreh::moreh_clip_grad_norm {
+
+inline uint32_t get_num_device_cores(Device *device) {
+    const auto num_cores_x = static_cast<uint32_t>(device->compute_with_storage_grid_size().x);
+    const auto num_cores_y = static_cast<uint32_t>(device->compute_with_storage_grid_size().y);
+    return num_cores_x * num_cores_y;
+}
+
+Tensor MorehClipGradNorm::invoke(
+    const std::vector<Tensor> &inputs,
+    float max_norm,
+    float norm_type,
+    bool error_if_nonfinite,
+    const std::optional<const Tensor> &total_norm,
+    const std::optional<MemoryConfig> &memory_config,
+    const std::optional<DeviceComputeKernelConfig> &compute_kernel_config) {
+    auto device = inputs.at(0).device();
+    const auto compute_kernel_config_val =
+        init_device_compute_kernel_config(device->arch(), compute_kernel_config, MathFidelity::HiFi4);
+
+    // Loop variable
+    const auto max_num_inputs = get_num_device_cores(device);
+    const auto total_num_inputs = static_cast<uint32_t>(inputs.size());
+    const auto num_iter = (total_num_inputs + max_num_inputs - 1) / max_num_inputs;
+
+    const auto &tmp_pow_sum = create_device_tensor(
+        SimpleShape{tt::constants::TILE_HEIGHT, tt::constants::TILE_WIDTH * static_cast<uint32_t>(inputs.size())},
+        inputs.at(0).get_dtype(),
+        Layout::TILE,
+        device,
+        memory_config.value_or(inputs.at(0).memory_config()));
+
+    // Run Step 1
+    uint32_t tile_offset{0};
+    auto num_inputs = total_num_inputs;
+    for (uint32_t i = 0; i < num_iter; i++) {
+        const auto num_inputs_at_this_iter = std::min(num_inputs, max_num_inputs);
+
+        ttnn::prim::moreh_clip_grad_norm_step1(
+            std::vector<Tensor>(inputs.begin() + tile_offset, inputs.begin() + tile_offset + num_inputs_at_this_iter),
+            norm_type,
+            tile_offset,
+            tmp_pow_sum,
+            memory_config,
+            compute_kernel_config_val);
+
+        if (i < (num_iter - 1)) {
+            tile_offset += num_inputs_at_this_iter;
+            num_inputs -= num_inputs_at_this_iter;
+        }
+    }
+
+    // Run Step 2
+    const auto &output_total_norm = ttnn::prim::moreh_clip_grad_norm_step2(
+        tmp_pow_sum,
+        norm_type,
+        total_norm,
+        memory_config,
+        init_device_compute_kernel_config(inputs.at(0).device()->arch(), compute_kernel_config, MathFidelity::HiFi4));
+
+    if (error_if_nonfinite) {
+        const auto fp32_total_norm =
+            tensor_impl::cast_vec<float>(owned_buffer::get_as<bfloat16>(output_total_norm.cpu())).at(0);
+        TT_ASSERT(
```

Comment:
```suggestion
        TT_FATAL(
```
Please, use TT_FATAL

Reply:
Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)